### ISSUE TITLE
Edge case when testing against local database

### ISSUE THREAD

**Issue by ElliotSalisbury** [django-salesforce/django-salesforce#231]
Hi,

Firstly, I love the work done on this, it's a really powerful tool.

I'm running into a problem when testing a salesforce database locally.
I believe the tables in the local database has been generated with an integer type for the ID field

For example:
```
a = Account(id="abc123abc123abc", ...)
a.save()
```
Would fail, as id has a type mismatch, and is expecting an integer

I believe this is because the `SalesforceAutoField`, as given to the id field in the `SalesforceModel` has not overriden the `get_internal_type()` method, and is defaulting to the integer type, of the parent class `AutoField`

Doing the following fixed my errors, but I wanted to check with you all first to see if I'm doing something wrong here. or if this should be pull requested?

```
class SalesforceAutoField(fields.AutoField):
    """
    An AutoField that works with Salesforce primary keys.

    It is used by SalesforceModel as a custom primary key. It doesn't convert
    its value to int.
    """
    description = _("Text")

    def get_internal_type(self):
        return "TextField"

    default_error_messages = {
        'invalid': _('This value must be a valid Salesforce ID.'),
    }

    def to_python(self, value):
        ...
```

Thanks
Elliot

**Comment by philchristensen** [comment]
So, you wouldn’t ever really set that parameter in normal usage, given that it’s an auto-generated field.

Are you attempting to essentially pre-seed the DB with objects that have deterministic primary keys? It’s possible that we’ve never tested in exactly this fashion, so I’ll have to defer to @hynekcer as to whether this is a bug or a known limitation.

When you apply your patch, does it pass the django-salesforce test suite?

**Comment by ElliotSalisbury** [comment]
I was writing some tests using a Factory to populate a local version of the salesforce db, something like this:

```
class SFAccountFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = 'eiq_salesforce.Account'

    name = Faker('company')
    description = Faker('text')
    website = Faker('url')
```

But I have some code elsewhere that allows users to search for SF objects by ID, and it ensures that the ID is either 15 or 18 characters long, which these autogenerated IDs were not. So it was raising errors, and my tests could not pass.

So then I added an id field to the factory, to make it for sure 18 chars long
```
def get_standardized_sf_id(sf_id):
    # ensure ID is salesforce 18 char case insensitive id
    # https://help.salesforce.com/articleView?id=000274977&type=1&language=en_US

    suffix = ""
    for i in range(3):
        loop = 0
        for position in range(5):
            current = sf_id[i * 5 + position]
            if current.isupper():
                loop += 1 << position
        suffix += "ABCDEFGHIJKLMNOPQRSTUVWXYZ012345"[loop]
    sf_id = sf_id[:15] + suffix
    return sf_id


class SFAccountFactory(factory.django.DjangoModelFactory):
    id = factory.Sequence(lambda n: get_standardized_sf_id('%015d' % n))
    ...
```

This then started throwing errors because it was trying to assign a string to an integer field in the database.
Hence my fix above.

About the test results, I haven't actually forked the project and run the test suite yet, so I don't know yet

**Comment by philchristensen** [comment]
If you’ve got a sandbox or dev account to test with, the unit tests are pretty thorough. Conversely, if you submit a PR, I can push it up to Travis for testing in our dev account.

I’ll still want to get @hynekcer’s input, but there’s a good chance we just haven’t run into this use case yet.

**Comment by ElliotSalisbury** [comment]
I've created the pull request (#232), but was unable to run the tests myself, so I was hoping you could use Travis, if that would be okay?

Atleast in my app it's been working out fine, but I have no idea how this kind of thing might effect the grander scheme of things.

**Comment by philchristensen** [comment]
Yep, I’ll run them on Travis. I’m out of pocket at the moment, but can get that going tonight.

-Phil
sent from my phone

**Comment by hynekcer** [comment]
I accept that it is not easy to create an exact copy of Salesforce data.
I accept also that the documentation about testing on "default" database should be better and added some tests (explained below in TODO list) 

The main missing comment is that **low level Salesforce database functionality can not be tested on a default non-salesforce backend**. I will explain why:

The most important feature of every database is that a unique primary key value is automatically assigned by the database (to prevent a race condition)  if no value is specified. That would be broken if `get_internal_type()` will return `"TextField"`. Even that PostgreSQL database supports also UUID primary key fields created by SQL `... id uuid PRIMARY KEY DEFAULT uuid_generate_v4()`, it is not supported by Django and UUID is not Text and is not 15/18 characters long etc. Many databases support even less then PostgreSQL.

Salesforce even doesn't allow to specify a primary key for a new object. If you write `a = Account(id="abc123abc123abc", ...)` then this id must exist yet and you must save it by `a.save(force_update=True)`. A more usual way is that you get an object from Salesforce, update it without touching the id and save it by a normal `save()`.

Your patch would break these automatically created primary keys. You would need for it a customized Salesforce model with a customized AutoField and probably also a customized Postgres backend. It is better to split your tests to a bigger more abstract universal part and a smaller part that works only with SFDC.

---

I think that the method `get_internal_type()` is intended for user provided custom fields that are based on a standard field. `AutoField` is a standard field and its `get_internal_type()` returns itself name (see `django/db/models/fields/__init__.py`).

``` python
def get_internal_type(self):
        return "AutoField"
```
Django-salesforce doesn't need to customize the method `AutoField.rel_db_type(connection)` because the "salesforce" backend is never migrated, but your patch it requires because that must be customized for other connections. You probably should also customize `db_type(connection)` because it the final type depends on the backend

---

TODO:

- fix README or also add something to ẅiki.
- add a test of basic supported functionality for testing of  applications with SalesforceModel used with a default (local) database backend assigned by SALESFORCE_DB_ALIAS
- Explore that (how) the primary key of Salesforce model is customizable.

**Comment by philchristensen** [comment]
Thanks for these insights, @hynekcer; this all makes sense to me.

It makes more sense to change the testing strategy than spending a lot of time on trying to make DB keys look like Salesforce IDs.

**Comment by thatguysimon** [comment]
Hey @ElliotSalisbury, here's what I came up with to tackle a similar problem:

I inherit from `SalesforceModel` and override the `id` field if I detect that we are in a test environment. I replace it with a `CharField` which is auto-populated when an object is created using a designated function.

Then when I run `makemigrations` and `migrate`, the tables in my testing sqlite3 database are created with a varchar id field, automatically populated, as expected. While the default functionality against SFDC remains the as is.

```python
# Creates a random 18 character string
def default_char_id():
    return ''.join(random.choices(string.ascii_lowercase + string.digits, k=18))

class TestableSalesforceModel(models.SalesforceModel):
    class Meta:
        abstract = True

    # Overriding the id field if we are in a test environment
    if settings.ENV == 'test':
        id = models.CharField(primary_key=True, max_length=18, name=fields.SF_PK, 
                              db_column='Id', verbose_name='ID', default=default_char_id)

class Account(TestableSalesforceModel):
    # Define your fields regularly...


# The following will create a new record in the Account table 
# with an auto-generated 18 character string id.
Account.objects.create() 
```

@philchristensen @hynekcer Thoughts?

**Comment by ElliotSalisbury** [comment]
@thatguysimon Thanks for your input! :)

In the end, I came up with a similar solution. Detecting when a test is running, and modifying the ID field type.
I use the pytest_configure function in order to do this.

So within my tests folder, I have a conftest.py that looks like this:
```
import django

def pytest_configure():
    django.setup()

    # override a function in the ID field of the django-salesforce models
    # we only want to do this in a test runner, to avoid any issues with accessing the live salesforce database
    # the problem is that the autofield creates an 'int' id column, but salesforce id's should be a 'char' field
    # this isnt a problem in the live database, because its maintained by SF, but local copies used for testing
    # are generated incorrectly.
    from salesforce.fields import SalesforceAutoField
    SalesforceAutoField.get_internal_type = lambda s: "TextField"
```

In your solution, although this seems incredibly unlikely, and you might never run into it, it might be possible in your solution to generate 2 Account objects with the same ID, given that the ID is generated randomly.
To avoid something like that, I generate my Accounts with a factory using a sequence:

```
def get_standardized_sf_id(sf_id):
    # ensure ID is salesforce 18 char case insensitive id
    # https://help.salesforce.com/articleView?id=000274977&type=1&language=en_US

    suffix = ""
    for i in range(3):
        loop = 0
        for position in range(5):
            current = sf_id[i * 5 + position]
            if current.isupper():
                loop += 1 << position
        suffix += "ABCDEFGHIJKLMNOPQRSTUVWXYZ012345"[loop]
    sf_id = sf_id[:15] + suffix
    return sf_id


class SFBaseFactory(factory.django.DjangoModelFactory):
    class Meta:
        abstract = True

    # creates id's that are incrementing integers, padded with 15 0's, and concatenated with the salesforce case insentive code.
    # e.g., "000000000000001AAA"
    id = factory.Sequence(lambda n: get_standardized_sf_id('%015d' % n))
    last_modified_date = factory.LazyFunction(timezone.now)
    ...
```

**Comment by hynekcer** [comment]
I describe requirements for a good solution (that I tried to implement):

* It should automatically assign primary keys to new unsaved objects.
* It should support copying objects from Salesforce database to an alternate database by `save(using='...')` (Then it can also easily read objects from a Salesforce backup in a zip csv format and save it do an alternate database).
* An example of alternative primary keys in Django docs are *Universally unique identifiers*  [UUIDField](https://docs.djangoproject.com/en/2.2/ref/models/fields/#s-uuidfield) with `default=uuid.uuid4`. That function uuid4() returns 128-bit values that are unique enough even with multi-thread, multi-process or distributed deployments.
* I remember some old documentation from Salesforce that the internal structure of primary key or its size (15/18) can change any time and the code should not depend on it. I have a similar function `case_safe_sf_id(id_15)` in `tests.test_mock.mocksf` module for tests of the framework, but I will never put any dependency on 15/18 characters to a production code. A `varchar(32)` is acceptable for UUID and also for Salesforce ID.
* I prefer writing the new code into a new module with another class SalesforceModel, e.g. sf_alternate_models.py to be possible to easily switch the stable standard implementation and the extended. I expect some issue reports. That should be avoided in production.

I tried two unsuccessful ways to implement this. It didn't work eithe in 'salesforce' or in 'default'. My current resume is that no solution exists by modifying only the the class for primary key and that the method `.save()` must be also overridden.

@ElliotSalisbury: If I understand, your problem and solution is related to `factory` in [factory_boy](https://github.com/FactoryBoy/factory_boy) package. That is too special. I a solution with requirements above will exist for uuid then it can be easily customized also to your case or maybe it can work out of box.)

Do you know a case where it is really necessary to use a 15/18 characters ID and to convert between them? They are interchangeable in API and SOQL. I think only about an equality check and check that an ID is present in a set of IDs or in a dictionary, where the size must be unified.

**Comment by hynekcer** [comment]
That solution is in PR #237
 
Changelog:
* Added: salesforce.models_extend module with SalesforceModel with
varchar primary key that works also with default databases.
Fixed methods for it: save() and bulk_create() to can create a new
pk or to copy an object exactly. Fixed #231 (4d042e5, 79603b7) 

The generator function for *salesforce alternate primary key* (default is ``uuid.uuid4().hex``) is configurable by ``settings.SF_ALT_PK``. This can be useful for your case if you want everything 15+3 chars.

Please try the new code.
